Yubico.YubiKey/src/Yubico/YubiKey/YubiKeyDeviceListener.cs

```diff
@@ -349,23 +351,46 @@ private static IEnumerable<IDevice> GetSmartCardDevices()
         private static void ErrorHandler(Exception exception) =>
             Log.GetLogger().LogWarning($"Exception caught: {exception}");
 
-        #region IDisposable Support
-
-        private bool _disposedValue;
+        private bool _isDisposed;
 
         /// <summary>
         /// Disposes the objects.
         /// </summary>
         /// <param name="disposing"></param>
         protected virtual void Dispose(bool disposing)
         {
-            if (!_disposedValue)
+            if (!_isDisposed)
             {
                 if (disposing)
                 {
-                    RwLock.Dispose();
+                    // Signal the listen thread that it's time to end.
+                    _tokenSource.Cancel();
+
+                    // Shut down the listener handlers.
```

Comment:
@GregDomzalski I'm not sure I see what you mean. We're not just stopping the thread, we're disposing the `YubiKeyDeviceListener` instance. Its `Dispose` method unsubscribes from the HID and smartcard listeners, and if they implement `IDisposable`, disposes of them. The references to them die when GC collects the YubiKeyDeviceListener instance. The finalizers all call `StopListening` in one way or the other.

I suppose one step further we could go is to make all of the listeners IDisposable (the HID ones are not currently) so that we can hasten their demise, and leave the finalizer so that legacy code doesn't get broken. I don't really think this is necessary, though.

Reply:
Hmm yeah. I see what you mean. I must have looked at the HID listeners when originally making this comment. It seems very strange to me that we chose to implement a finalizer only and not `IDisposable`. That doesn't really seem very kosher. But it is what it is, and I agree - I believe your code is doing the right thing based on what's present elsewhere. Feel free to resolve this thread.

Reply:
@DennisDyallo you good with merging to develop?